lib/can-npm-publish.js

```diff
@@ -69,32 +69,82 @@ const viewPackage = (packageName, registry) => {
     return new Promise((resolve, reject) => {
         const registryArgs = registry ? ["--registry", registry] : [];
         const view = spawn("npm", ["view", packageName, "versions", "--json"].concat(registryArgs));
-        let result = "";
-        let errorResult = "";
+        let _stdoutResult = "";
+        let _stderrResult = "";
 
+        const getJsonOutputs = ({ stdout, stderr }) => {
+            let stdoutJSON = {};
+            let stderrJSON = {};
+            try {
+                stdoutJSON = JSON.parse(stdout);
+            } catch (error) {
+                // nope
+                // console.warn("stdoutJSON parse error", stdout);
+            }
+            try {
+                stderrJSON = JSON.parse(stderr);
+            } catch (error) {
+                // nope
+                // console.warn("stderrJSON parse error", stderr);
+            }
+            return {
+                stdoutJSON,
+                stderrJSON
+            };
+        };
+        const isError = (json) => {
+            return json && "error" in json;
+        };
+        const is404Json = (json) => {
+            return isError(json) && json.error.code === "E404";
+        };
         view.stdout.on("data", (data) => {
-            result += data.toString();
+            _stdoutResult += data.toString();
         });
 
         view.stderr.on("data", (err) => {
-            errorResult += err.toString();
+            const stdErrorStr = err.toString();
+            try {
+                // Workaround for npm 7
+                // npm 7 --json option is broken
+                // It aim to remove non json output.
+                // FIXME: However,This logics will break json chunk(chunk may be invalid json)
+                // https://github.com/azu/can-npm-publish/issues/19
+                // https://github.com/npm/cli/issues/2740
+                JSON.parse(stdErrorStr); // parse check
```

Comment:
npm 7 workaround

- [[BUG] npm view --json sompackage dist-tags has invalid json · Issue #3075 · npm/cli](https://github.com/npm/cli/issues/3075)
- [[BUG] npm view --json outputs illegal JSON for existing package but non-existent version · Issue #3184 · npm/cli](https://github.com/npm/cli/issues/3184)
- [[BUG] NPM 7.x broke the "--json" CLI parameter · Issue #2740 · npm/cli](https://github.com/npm/cli/issues/2740)

Reply:
npm 7 outputs
```
"npm"
"Err!"
""
"JSON strings..."
""
```
...

Reply:
This logic will break following pattern

```
"npm"
"Err!"
""
"JSON part1"
"JSON part2"
""
```

Reply:
use https://github.com/bconnorwhite/extract-first-json for the error
https://github.com/azu/can-npm-publish/runs/2900472385